### PR TITLE
Feature/declarative spec

### PR DESCRIPTION
**Pull request by sebastian-echeverria**
This is the second attempt to implement more declarative specifications. It is a fully working PR, but it does not cover all functionality that will be added for this. It covers only half of the discussion detailed in #135, and a second PR will be sent later to address the remaining points.

The following are new changes included in this second attempt:
- Name changes: Result was renamed to Value, and ValidationResult was renamed to Result.
- A label was added to Conditions, to keep track of them and more easily handle automatic validation.
- A rationale field was added to Properties so that the details on why a property is important for a spec can be captured.
- Backwards compatibility was removed, now a Spec has to define Conditions (if any) when it is constructed, upfront.
- Binding was removed from the code and is no longer supported, having been replaced with Conditions and SpecValidator.
- Schemas and docs were updated.
- Issues #98 and #99 are addressed. Value was moved from inside measurements package to a new "value" package, and the newly renamed Result was moved to "validation". Also, measurement_metadata was moved to a first-level package, and now is used directly by both Value and Result (ex ValidationResult), instead of Result embedding a Value, and instead of Value deconstructing MeasurementMetadata into a separate measurement type and identifier fields. This also avoids previous circular dependencies between classes that used to be in the "measurement" package. Now all 3 classes, Measurement, Value and Result, use MeasurementMetadata as the way to maintain the details/keep track of the measurement type and id of the measurement being done. Package name may need improvement though.
 
Main changes kept from the first attempt (from temporary pull request 131):
- New way to specify Spec details was added. The concept of Condition was created and can be associated to a Spec. Conditions are composed of a label, a measurement type, a validator (method name), and a threshold/param for the validator. They are added for a Property in the spec, and a Property can have multiple conditions. In the flow, these will be added at the when creating a Spec, along with the properties. This means that a Specification is not only a list of properties, but also a list of how to validate each property. (NOTE: Their name will likely change in the next PR). 
- A new class, SpecValidator, was added. Values can be added to this class, along with the Spec, and it automatically validates the spec by calling the configured validators in the conditions for the Spec, for each Value.
- The schema for Spec has been changed, and now it contains a list of conditions for each property.
- The schemas for the Spec and BoundSpec have been unified in practice, with both documents having the exact same structure, EXCEPT for an additional "validation" section for each condition that is only present in the BoundSpec. Basically, the BoundSpec is now a Spec with all conditions having a validation section with the Results (maybe more of a ValidatedSpec than a BoundSpec at this point). There is probably still work here on the best way to merge these two things. For example, it may make more sense for the class with the validations to be called ValidatedSpec and derive from Spec, than to be a separate class as BoundSpec is.

Items still missing from #135 that will be addressed in a separate PR:
 - Condition and Validator have not been renamed yet.
 - Approach of using thunks and auto completion when defining conditions/validators for a spec is not implemented yet.
 - Combination of conditions other than an implicit "and" is supported yet.
 - No .value property in Measurements yet.
 - No changes to integrate ExternalMeasurements as described yet.

### CODE REVIEW

File: src/mlte/spec/condition.py
Comment (turingcompl33t) [inline src/mlte/spec/condition.py:28]: Do we need to think of something more "rigorous" for specifying thresholds?

File: src/mlte/value/types/real.py
Comment (turingcompl33t) [inline src/mlte/value/types/real.py:143]: I believe you intend to address this in a separate PR, but we decided that these methods will return thunks rather than the result of evaluation itself, correct?

**Review by turingcompl33t — CHANGES_REQUESTED** [review]
I tried to focus in on just the major functional changes rather than the large diff that results from the name change. I think we have just a few points to talk about before we merge.

File: src/mlte/measurement/cpu/local_process_cpu_utilization.py
Comment (sebastian-echeverria) [reply src/mlte/measurement/cpu/local_process_cpu_utilization.py:13]: Totally agree. We can discuss this a bit more on our upcoming meeting.

File: src/mlte/measurement/measurement.py
Comment (sebastian-echeverria) [reply src/mlte/measurement/measurement.py:38]: Inside MeasurementMetadata, it will be an identifier. But instead of passing an identifier as an argument, I am passing a string, and only interally it will create an Identifier based on this. To be honest, I don't recall exactly why I made this change but I can check. It made sense to simplify some of the depdencies. I think it was to avoid having Measurement depend on Identifier as well, and only make it depend on MeasurementMetadata.

File: src/mlte/spec/condition.py
Comment (sebastian-echeverria) [reply src/mlte/spec/condition.py:17]: So I get this can be confusing, mainly because this PR is a "part 1" for the declarative spec, and won't be fully completed until the next PR. When I initially came up with the Condition class, it was an association of something you wanted to validate eventually, which included a measurement type, a validation method and arguments for that. In our last meeting, you suggested, as you are indicating above, to basically rename the Validator into Condition, and use some other names for this association (I think RegExp), and changing it to be a thunk. Don't worry, I got that. But as I mentioned in the PR, all of that is not included yet in this PR. I am still using my initial definition of Condition, which is not the condition you want, for this first step. I will rename the Validator to Condition and work with the thunks and everything in the next PR. Since there was so much renaming going on, and I already had a Condition defined in my old way here, I prefered to leave that to the second PR. We can discuss this at our upcoming meeting, and I understand if you would prefer for me to continue working on this PR instead of dividing it into two steps, if you are not confortable with this as an intermediate step.

File: src/mlte/spec/condition.py
Comment (sebastian-echeverria) [reply src/mlte/spec/condition.py:28]: Definitely, and that will come with the next PR and thunks and everything as we had discussed.

File: src/mlte/spec/spec.py
Comment (sebastian-echeverria) [reply src/mlte/spec/spec.py:21]: Starting from Python 3.9, "list" can be used as the type for type hints, instead of List. Same for dict and others. And to get this to work on 3.8, just adding "from __future__ import annotations" makes it work the same way. Since using list and dict instead of List and Dict is the way forward for Python type hints, I am trying to switch to that usage when possible. See https://stackoverflow.com/a/65777001

File: src/mlte/spec/spec.py
Comment (sebastian-echeverria) [reply src/mlte/spec/spec.py:59]: See answer above.

File: src/mlte/spec/spec.py
Comment (sebastian-echeverria) [reply src/mlte/spec/spec.py:61]: Yep, that was an outdated piece of comment, I'll remove it.

File: src/mlte/spec/spec_validator.py
Comment (sebastian-echeverria) [reply src/mlte/spec/spec_validator.py:19]: It is used directly. Spec is currently growing too much, and I think mixing the responsibilities of defining the spec and validating results for it in the same class is not a good idea. That being said, there are modifications that could be make, like keeping this SpecValidator but having Spec call it internally. We can discuss this in our meeting tomorrow.

File: src/mlte/value/types/real.py
Comment (sebastian-echeverria) [reply src/mlte/value/types/real.py:143]: Yes, that will be the case in the next PR.

File: src/mlte/spec/condition.py
Comment (turingcompl33t) [reply src/mlte/spec/condition.py:17]: Resolved during discussion.

File: src/mlte/measurement/measurement.py
Comment (turingcompl33t) [reply src/mlte/measurement/measurement.py:38]: Resolved.

File: src/mlte/spec/spec.py
Comment (turingcompl33t) [reply src/mlte/spec/spec.py:21]: Great! Moving us into the next generation.

File: src/mlte/spec/spec.py
Comment (turingcompl33t) [reply src/mlte/spec/spec.py:59]: Resolved.

File: src/mlte/spec/spec.py
Comment (turingcompl33t) [reply src/mlte/spec/spec.py]: Done.

File: src/mlte/spec/spec.py
Comment (turingcompl33t) [reply src/mlte/spec/spec.py:94]: Moving internally, removing from public interface for now.

File: src/mlte/spec/spec_validator.py
Comment (turingcompl33t) [reply src/mlte/spec/spec_validator.py:19]: We discussed this. I like Sebastian's plan of separating the `Spec` and the validation functionality.

File: src/mlte/value/types/real.py
Comment (turingcompl33t) [reply src/mlte/value/types/real.py:143]: Resolved.

**Review by turingcompl33t — APPROVED** [review]
Thanks @sebastian-echeverria. Approving and merging.

**Pull request merged**
